### PR TITLE
Add render functionality

### PR DESCRIPTION
**Pull request by chrismccord**

### CODE REVIEW

File: lib/phoenix/template.ex [hunk 1]
```diff
@@ -54,6 +54,130 @@ defmodule Phoenix.Template do
...
+  For a module called `MyApp.FooHTML` and template "index.html.heex",
+  it will:
+
+    * First attempt to call `MyApp.FooHTML.index(assigns)`
+
+    * Then fallback to `MyApp.FooHTML.render("index.html", assigns)`
+
+    * Raise otherwise
+
+  It expects the HTML module, the template as a string, and a
```
Comment (chrismccord) [inline lib/phoenix/template.ex]:
```suggestion
  It expects the HTML module, the template as a string or atom, the format, and a
```

File: lib/phoenix/template.ex [hunk 2]
```diff
@@ -54,6 +54,130 @@ defmodule Phoenix.Template do
...
+  Phoenix, they are:
+
+    * `:layout` - tells Phoenix to wrap the rendered result in the
+      given layout. See next section
+
+  ## Layouts
+
+  Templates can be rendered within other templates using the `:layout`
+  option. `:layout` accepts a tuple of the form
+  `{LayoutModule, "template.extension"} | {LayoutModule, :template}`.
```
Comment (josevalim) [inline lib/phoenix/template.ex]: We do not need to support the string format here. Let's go with only atoms for layout.

File: lib/phoenix/template.ex [hunk 3]
```diff
@@ -54,6 +54,130 @@ defmodule Phoenix.Template do
...
+       when is_atom(layout_mod) and (is_binary(layout_tpl) or is_atom(layout_tpl)) do
+    content = render_with_fallback(module, template, format, assigns)
+    assigns = Map.put(assigns, :inner_content, content)
+    render_with_fallback(layout_mod, layout_tpl, format, assigns)
+  end
+
+  defp render_within_layout({layout, _assigns}, _module, _template, _format) do
+    raise ArgumentError, """
+    invalid value for reserved key :layout in Phoenix.Template.render/4 assigns
+    :layout accepts a tuple of the form {LayoutModule, "template.extension"}
```
Comment (josevalim) [inline lib/phoenix/template.ex]:
```suggestion
    invalid value for reserved key :layout in Phoenix.Template.render/4 assigns.
    :layout accepts a tuple of the form {LayoutModule, "template.extension"},
```

File: lib/phoenix/template.ex [hunk 4]
```diff
@@ -54,6 +54,130 @@ defmodule Phoenix.Template do
     end
   end
 
+  @doc """
+  Renders the template and returns iodata.
+  """
+  def render_to_iodata(module, template, format, assign) do
+    module
+    |> render(template, format, assign)
+    |> encode(template_with_ext(template, format))
```
Comment (josevalim) [inline lib/phoenix/template.ex]: Let's only pass the format to `encode` and change Phoenix.View to call `Path.extname`.

File: lib/phoenix/template.ex [hunk 5]
```diff
@@ -54,6 +54,130 @@ defmodule Phoenix.Template do
...
+    if encoder = format_encoder(template_with_ext) do
+      encoder.encode_to_iodata!(content)
+    else
+      content
+    end
+  end
+
+  defp render_with_fallback(module, template, format, assigns)
+       when is_atom(module) and is_atom(template) and is_binary(format) and is_map(assigns) do
+    if function_exported?(module, template, 1) do
```
Comment (josevalim) [inline lib/phoenix/template.ex]: We need the erlang module_loaded check both before String.to_existing_atom and the function_exported? check, so we should probably move it somewhere so we can check it only once:

```suggestion
    :erlang.module_loaded(module) or :code.ensure_loaded(module)

    if function_exported?(module, template, 1) do
```

File: lib/phoenix/template/eex_engine.ex [hunk 6]
```diff
@@ -10,7 +10,9 @@ defmodule Phoenix.Template.EExEngine do
   end
 
   defp options_for(path) do
-    case Phoenix.Template.format_encoder(Path.rootname(path)) do
+    "." <> format = path |> Path.rootname() |> Path.extname()
```
Comment (josevalim) [inline lib/phoenix/template/eex_engine.ex:13]: You don’t need rootname I think.

File: lib/phoenix/template.ex [hunk 7]
```diff
@@ -121,9 +121,10 @@ defmodule Phoenix.Template do
 
   defp render_within_layout({{layout_mod, layout_tpl}, assigns}, module, template, format)
        when is_atom(layout_mod) and is_binary(layout_tpl) do
+    layout_base = Path.rootname(layout_tpl)
```
Comment (josevalim) [inline lib/phoenix/template.ex]: The layout template should be without the format?

**Review by josevalim — APPROVED** [review]
Two comments and ship it!

**Comment by chrismccord** [comment]
❤️❤️❤️🐥🔥

**Pull request merged**